Feat: Add Literal AI Integration

willydouhard opened the pull request:
## Title

Add Literal AI integration to Lite LLM

## Relevant issues

## Type

🆕 New Feature


## Changes

- Updating the readme to mention Literal AI in the logging section
- Adding the `LiteralAILogger` custom logger class
- Updating [litellm_logging.py](https://github.com/BerriAI/litellm/compare/main...Chainlit:litellm:main#diff-202545f3ab4f290ac9967d9ba19689aafe15902296a2ef5a6ca5410a93eaa38b) to account for `LiteralAILogger`
- Adding a dedicated documentation page


## [REQUIRED] Testing - Attach a screenshot of any new tests passing local
No test added.

willydouhard committed:
feat: add Literal AI integration

willydouhard committed:
Merge branch 'main' of github.com:Chainlit/litellm

willydouhard committed:
Merge branch 'main' of github.com:BerriAI/litellm

willydouhard commented:
Fixed conflicts with the main branch.

willydouhard committed:
Merge branch 'main' of github.com:BerriAI/litellm

willydouhard committed:
update readme

willydouhard committed:
Update README.md

krrishdholakia commented on litellm/integrations/literal_ai.py:
```diff
@@ -0,0 +1,193 @@
+from litellm.litellm_core_utils.redact_messages import redact_user_api_key_info
+
+
+class LiteralAILogger:
```
@willydouhard could you please use the custombatchlogger here? 

https://github.com/BerriAI/litellm/blob/ab153c3e8dab6be093eba3726ada19da176ee8cc/litellm/integrations/langsmith.py#L66

it would prevent issues during high traffic

krrishdholakia commented on litellm/integrations/literal_ai.py:
```diff
@@ -0,0 +1,193 @@
+                f"\033[91mLiteralAI not installed, try running 'pip install literalai' to fix this error: {e}\n{traceback.format_exc()}\033[0m"
+            )
+    
+        self.client = LiteralClient(api_key=literalai_api_key, url=literalai_api_url)
```
can we not require the literalai client here? 

It will be harder for proxy users to use this, as it would need to be in the docker image. 

Langsmith is a good reference implementation that uses httpx calls - https://github.com/BerriAI/litellm/blob/ab153c3e8dab6be093eba3726ada19da176ee8cc/litellm/integrations/langsmith.py#L66

krrishdholakia commented on litellm/integrations/literal_ai.py:
```diff
@@ -0,0 +1,193 @@
+
+       
+    @staticmethod
+    def add_metadata_from_header(litellm_params: dict, metadata: dict) -> dict:
```
This shouldn't be necessary if you're using our standard logging payload - https://github.com/BerriAI/litellm/blob/ab153c3e8dab6be093eba3726ada19da176ee8cc/litellm/integrations/gcs_bucket.py#L63

It should guarantee the metadata, etc. is correctly handled for all the different llm api calls (audio, embedding, chat, etc.) as well as if users set flags for masking certain information.

krrishdholakia commented on litellm/integrations/literal_ai.py:
```diff
@@ -0,0 +1,193 @@
+                    }
+                }
+            if response_obj is not None and response_obj.get("id", None) is not None:
+                generation_id = litellm.utils.get_logging_id(start_time, response_obj)
```
this should be handled by the standard logging payload id

krrishdholakia commented on litellm/integrations/literal_ai.py:
```diff
@@ -0,0 +1,193 @@
+            if response_obj is not None and response_obj.get("id", None) is not None:
+                generation_id = litellm.utils.get_logging_id(start_time, response_obj)
+                step["metadata"]["litellm_id"] = generation_id
+                step["generation"]["inputTokenCount"] = response_obj.usage.prompt_tokens
```
not every llm response contains a usage object (e.g. rerank and `/audio/speech`) endpoints. 

again, would prefer using the standard logging payload's usage information fields here - https://github.com/BerriAI/litellm/blob/ab153c3e8dab6be093eba3726ada19da176ee8cc/litellm/types/utils.py#L1314

krrishdholakia commented on litellm/integrations/literal_ai.py:
```diff
@@ -0,0 +1,193 @@
+                step["generation"]["outputTokenCount"] = response_obj.usage.completion_tokens
+                step["generation"]["tokenCount"] = response_obj.usage.prompt_tokens + response_obj.usage.completion_tokens
+
+            if (
```
this shouldn't be handled here anymore, as it will require us to update each integration when we add support for different endpoints. 

Use the standard logging payload's `messages` and `response` fields instead - https://github.com/BerriAI/litellm/blob/ab153c3e8dab6be093eba3726ada19da176ee8cc/litellm/types/utils.py#L1331

krrishdholakia commented on litellm/litellm_core_utils/litellm_logging.py:
```diff
@@ -433,6 +435,22 @@ def pre_call(self, input, api_key, model=None, additional_args={}):
                             message=f"Model Call Details pre-call: {details_to_log}",
                             level="info",
                         )
+                    elif callback == "literalai":
```
if the LiteralAILogger inherits from CustomLogger (or CustomBatchLogger) this should not be necessary. 

All you need to do is add it here - https://github.com/BerriAI/litellm/blob/ab153c3e8dab6be093eba3726ada19da176ee8cc/litellm/litellm_core_utils/litellm_logging.py#L2137

krrishdholakia commented on litellm/litellm_core_utils/litellm_logging.py:
```diff
@@ -900,6 +918,43 @@ def success_handler(
                             end_time=end_time,
                             print_verbose=print_verbose,
                         )
+                    if callback == "literalai":
```
same comment here - this should be avoided. 

as long as this is a CustomLogger class, we know how to log to it. 

https://github.com/BerriAI/litellm/blob/ab153c3e8dab6be093eba3726ada19da176ee8cc/litellm/litellm_core_utils/litellm_logging.py#L2137

krrishdholakia commented:
Hey @willydouhard summarizing feedback here: 

- refactor LiteralAILogger to inherit from [CustomBatchLogger](https://github.com/BerriAI/litellm/blob/ab153c3e8dab6be093eba3726ada19da176ee8cc/litellm/integrations/custom_batch_logger.py#L18). [Reference Langsmith implementation](https://github.com/BerriAI/litellm/blob/ab153c3e8dab6be093eba3726ada19da176ee8cc/litellm/integrations/langsmith.py#L66)
- move from LiteralAIClient to httpx client for proxy user ease of use. [Reference Langsmith Implementation](https://github.com/BerriAI/litellm/blob/ab153c3e8dab6be093eba3726ada19da176ee8cc/litellm/integrations/langsmith.py#L66)
- move to reading information from standardloggingpayload (metadata, usage information, request/response) https://github.com/BerriAI/litellm/blob/ab153c3e8dab6be093eba3726ada19da176ee8cc/litellm/integrations/gcs_bucket.py#L63

willydouhard committed:
Merge branch 'main' of github.com:BerriAI/litellm

willydouhard committed:
fix: address comments

willydouhard committed:
Merge branch 'main' of github.com:Chainlit/litellm

willydouhard commented:
> Hey @willydouhard summarizing feedback here:
> 
> * refactor LiteralAILogger to inherit from [CustomBatchLogger](https://github.com/BerriAI/litellm/blob/ab153c3e8dab6be093eba3726ada19da176ee8cc/litellm/integrations/custom_batch_logger.py#L18). [Reference Langsmith implementation](https://github.com/BerriAI/litellm/blob/ab153c3e8dab6be093eba3726ada19da176ee8cc/litellm/integrations/langsmith.py#L66)
> * move from LiteralAIClient to httpx client for proxy user ease of use. [Reference Langsmith Implementation](https://github.com/BerriAI/litellm/blob/ab153c3e8dab6be093eba3726ada19da176ee8cc/litellm/integrations/langsmith.py#L66)
> * move to reading information from standardloggingpayload (metadata, usage information, request/response) https://github.com/BerriAI/litellm/blob/ab153c3e8dab6be093eba3726ada19da176ee8cc/litellm/integrations/gcs_bucket.py#L63

Thank you for your review @krrishdholakia. I still have to check a couple things on my end but is this what you had in mind?

willydouhard committed:
fix: remove literalai sdk

krrishdholakia commented on litellm/integrations/literal_ai.py:
```diff
@@ -0,0 +1,348 @@
+        variables = self._steps_variables_builder(self.log_queue)
+
+        try:
+            response = requests.post(
```
use the HTTPHandler here https://github.com/BerriAI/litellm/blob/25edb4ed652f10c4048ff6e6ae54812ea7506d2a/litellm/llms/custom_httpx/http_handler.py#L283 

it'll handle any weird ssl issues

krrishdholakia commented on litellm/integrations/literal_ai.py:
```diff
@@ -0,0 +1,348 @@
+            
+            prompt_id = None
+            variables = None
+            if "orig_messages" in clean_metadata:
```
if you want to grab anything passed in from the client request, i think it's fine to just do kwargs.get("metadata") 

just don't log the kwargs["metadata"] to the destination, as it leads to: 
1. failed logs if metadata values can't be serialized 
2. potential information leaks

krrishdholakia commented:
Looks much better. 

1. Can you add a couple unit tests to make sure things work as expected + no future regressions - https://github.com/BerriAI/litellm/blob/25edb4ed652f10c4048ff6e6ae54812ea7506d2a/litellm/tests/test_langsmith.py#L26
2. Just check you're able to get all your needed fields from metadata (see suggestion on kwargs vs. standard payload)

once you're done, if you can share a screenshot of it passing your tests i think we should be good to merge

willydouhard committed:
fix: use HTTPHandler

willydouhard committed:
chore: add test

willydouhard commented:
I think I addressed all of the comments. Can you point me to some documentation on how to run the tests? I see the `python3 -m pytest` command but it asks for a lot of API keys.

willydouhard committed:
fix: add asyncio lock

willydouhard commented:
<img width="1534" alt="Screenshot 2024-09-30 at 10 05 15" src="https://github.com/user-attachments/assets/8fe8fe2f-416c-4beb-a65c-14ba5b762bef">
Here is a screenshot for the test I added specifically.

krrishdholakia commented on docs/my-website/docs/observability/literalai_integration.md:
````diff
@@ -0,0 +1,119 @@
+Ensure you have the `literalai` package installed:
+
+```shell
+pip install literalai litellm
````
i'm assuming this is no longer needed

willydouhard replied:
Correct
